source/agora/test/EnrollmentManager.d

```diff
@@ -167,3 +196,235 @@ unittest
+        if (next.enrolls.length > 0)
+            next.enrolls.length = 0;
+
+        auto prev_value = prev.serializeFull().toVec();
+        auto next_value = next.serializeFull().toVec();
+        if (this.scp.nominate(slot_idx, next_value, prev_value))
+        {
+            log.info("{}(): Tx set nominated", __FUNCTION__);
+        }
+        else
+        {
+            log.info("{}(): Tx set rejected nomination", __FUNCTION__);
+        }
+    }
+}
+
+class BynzantineNode : TestNode
+{
+    this (Config config, Registry* reg)
+    {
+        super(config, reg);
+    }
+
+    override Nominator getNominator (KeyPair key_pair, Ledger ledger,
+        TaskManager taskman, NetworkClient[PublicKey] quorum_peers,
+        SCPQuorumSet quorum_set)
+    {
+        return new ByzantineNominator(key_pair, ledger, taskman,
+            quorum_peers, quorum_set);
+    }
+}
+
+class ByzantineManager : TestAPIManager
+{
+    override void createNewNode (PublicKey address, Config conf)
+    {
+        RemoteAPI!TestAPI api;
+        if (this.nodes.length == 0)  // first node is byzantine
+            api = RemoteAPI!TestAPI.spawn!(BynzantineNode)(conf, &this.reg);
+        else
+            api = RemoteAPI!TestAPI.spawn!(TestNode)(conf, &this.reg);
+
+        this.reg.register(address.toString(), api.tid());
+        this.nodes ~= NodePair(address, api);
+    }
+}
+
+/// 'Four nodes enrollment' Nomal test
+unittest
+{
+    import std.algorithm;
+    import std.conv;
+    import core.time;
+
+    TestConf conf = { nodes : 4, threshold : 4 };
+    auto network = makeTestNetwork(TestConf.init);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto nodes = network.clients;
+    auto node_1 = nodes[$ - 1];
+
+    auto gen_key_pair = getGenesisKeyPair();
+
+    Transaction[] txs;
+    Transaction payment_tx;
+    foreach (idx; 0 .. Block.TxsInBlock)
+    {
+        Transaction tx;
+        Output output;
+
+        tx.inputs ~= Input(hashFull(GenesisTransaction), idx.to!uint);
+        if (idx < nodes.length)
+        {
+            tx.type = TxType.Freeze;
+            tx.outputs ~= Output(Amount.MinFreezeAmount, nodes[idx].getPublicKey);
+        }
+        else
+        {
+            tx.type = TxType.Payment;
+            foreach (idx2; 0 .. Block.TxsInBlock)
+                tx.outputs ~= Output(Amount(10_000_000), gen_key_pair.address);
+
+            if (payment_tx == Transaction.init)
+                payment_tx = tx;
+        }
+
+        auto signature = gen_key_pair.secret.sign(hashFull(tx)[]);
+        tx.inputs[0].signature = signature;
+        txs ~= tx;
+    }
+
+    txs.each!(tx => node_1.putTransaction(tx));
+    containSameBlocks(nodes, 1).retryFor(8.seconds);
+
+    Enrollment[] enrolls;
+    foreach (node; nodes)
+    {
+        Enrollment enroll = node.createEnrollmentData();
```

Comment:
This API needs to be removed. We can't have an API asking node to enroll.

Reply:
It's already been made by @linked0  for the test.

Reply:
> This API needs to be removed. We can't have an API asking node to enroll.

We need a way to broadcast enrollment information which is implemented as `enrollValidator`. And, the `createEnrollmentData` was implemented only for integration test because there is no way to make EnrollmentManager create an enrollment information.